packages/editor/src/lib/utils/reorderShapes.ts

```diff
@@ -147,14 +149,49 @@ function reorderToFront(moving: Set<TLShape>, children: TLShape[], changes: TLSh
 	}
 }
 
+function getVerticesInPageSpace(editor: Editor, shape: TLShape) {
+	const geo = editor.getShapeGeometry(shape)
+	const pageTransform = editor.getShapePageTransform(shape)
+	if (!geo || !pageTransform) return null
+	return pageTransform.applyToPoints(geo.vertices)
+}
+
+function getOverlapChecker(editor: Editor, moving: Set<TLShape>) {
+	const movingVertices = Array.from(moving)
+		.map((shape) => {
+			const vertices = getVerticesInPageSpace(editor, shape)
+			if (!vertices) return null
+			return { shape, vertices }
+		})
+		.filter(Boolean) as { shape: TLShape; vertices: Vec[] }[]
+
+	const isOverlapping = (child: TLShape) => {
+		const vertices = getVerticesInPageSpace(editor, child)
+		if (!vertices) return false
+		return movingVertices.some((other) => {
+			return polygonsIntersect(other.vertices, vertices)
+		})
+	}
+
+	return isOverlapping
+}
+
 /**
  * Reorders the moving shapes forward in the parent's children.
  *
+ * @param editor The editor
  * @param moving The set of shapes that are moving
  * @param children The parent's children
  * @param changes The changes array to push changes to
  */
-function reorderForward(moving: Set<TLShape>, children: TLShape[], changes: TLShapePartial[]) {
+function reorderForward(
+	editor: Editor,
+	moving: Set<TLShape>,
+	children: TLShape[],
```

Comment:
I might move the filtering out of the function here and into whatever passes in the `children` (and maybe rename that argument to reflect this). There are times when reordering forward in a dumb way would be good (ie from a sidebar) so if we can keep this dumb then that would be best.

Reply:
check alex's comment above, it's not totally safe to do that because then we'd be relying on the jitter algorithm to avoid clashing indexes. But we could definitely make smartness opt-in.